### ISSUE TITLE
Smart BMS JK-PB2A16S-20P extra settings required

### ISSUE THREAD

**Issue by silviudc** [fl4p/batmon-ha#169]
Is the Smart BMS JK-PB2A16S-20P with Active Balancer supported by this integration? This is the new model that talks to inverters.
https://www.jkbms.com/product/jk-pb2a16s20p-jk-pb2a16s-20p/

**Comment by fl4p** [comment]
It might work. Let me know if you try it. There is no risk of harm.

**Comment by silviudc** [comment]
Just to confirm that it works but I see the following only:
![image](https://github.com/fl4p/batmon-ha/assets/17740184/0a1e282d-114a-4dda-ad8f-0662020a8a82)
Would it be possible to add/update the following:
1. Would like to see toggle buttons for balance and float mode, like for charge/discharge
2. Temp sensors 3 and 4 are missing, only 1 and 2 showing

**Comment by at906** [comment]
> Just to confirm that it works but I see the following only: ![image](https://private-user-images.githubusercontent.com/17740184/296725730-0a1e282d-114a-4dda-ad8f-0662020a8a82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwMTEwMzEsIm5iZiI6MTcwNjAxMDczMSwicGF0aCI6Ii8xNzc0MDE4NC8yOTY3MjU3MzAtMGExZTI4MmQtMTE0YS00ZGRhLWFkOGYtMDY2MjAyMGE4YTgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIzVDExNTIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3Njg1ZjU5Y2UxMDNjMGRkZTE2OGNlZTE1MGY1OTUzZTUzODJjMDE5NGRhZDgxNzBlY2QzYTNiYTY1YjU1NDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Xr0Ljz0QlGegOInYCtKRE_eW76H99eJJoNlGimeWCmg) Would it be possible to add/update the following:
> 
> 1. Would like to see toggle buttons for balance and float mode, like for charge/discharge
> 2. Temp sensors 3 and 4 are missing, only 1 and 2 showing
> 
> JKBMS FW is v14.17

Hello,

Can you walk me through this setup? MQTT Broker finds the inverter, but i am not able to populate any of the info

**Comment by fl4p** [comment]
I've added the other 2 temperature sensors.

@silviudc What is float mode? Can you post a screenshot of the app screen with that button?

**Comment by personshelldon** [comment]
@fl4p The same problem here. It would be perfect to have a "Float Mode" switch in Home Assistant. All the other settings work perfectly!

**Comment by personshelldon** [comment]
[float_charge.patch](https://github.com/user-attachments/files/16700404/float_charge.patch)

Maybe it will be useful for someone. Made a small patch that adds a "Float Charge" switch for the JKBMS PBx series. Personally, I need this switch for correct automation in Home Assistant. Use only at your own risk, since I do not know how this change will work for other JK BMS boards. My BMS is PB2A16S20P v15.24

**Comment by fl4p** [comment]
@personshelldon thanks for the patch.
I'd add this right away if we had some information which HW/SW version have this switch.

Also I dont fully understand float mode. This needs a regulator (buck converter) so the BMS can regulate charge current?

**Comment by roumano** [comment]
> [@personshelldon](https://github.com/personshelldon) thanks for the patch. I'd add this right away if we had some information which HW/SW version have this switch.
> 
> Also I dont fully understand float mode. This needs a regulator (buck converter) so the BMS can regulate charge current?

After the absorption mode, if the float mode is enable, the jk entering in the float mode (with a lower voltage value than the absorption mode to not stress the battery but continue to keep them fully charge

If the float mode is disable, it's say in absorption mode for ever (until the voltage decrease, so entering in bulk mode)

**Comment by Off-Grid-Garage** [comment]
> Also I dont fully understand float mode. This needs a regulator (buck converter) so the BMS can regulate charge current?

The JK has built-in Float Mode settings and timers. It requests a lower voltage from the inverter, so there is no additional hardware needed.

@roumano explained it already pretty good. Maybe these graphs help as well.

With Float Mode enabled:
![Image](https://github.com/user-attachments/assets/9b367596-f8c3-4b57-8edd-9caf711b6525)

With Float Mode disabled:
![Image](https://github.com/user-attachments/assets/c31764d7-bb9e-48ff-b3ae-136e597dad6c)

**Comment by fl4p** [comment]
@Off-Grid-Garage thanks for the explanation.
So the voltage drop we see when BMS enters float mode comes from the BMS? 
I guess it has an internal buck converter to step-down the voltage.

Does this work without talking to the solar charger?


Anyone knows which models has the float mode? 
looking at https://www.jkbms.com/products/active-balancer-bms/ , looks like the PB* models all have power inductors that could be part of the internal buck convert..

**Comment by fl4p** [comment]
I've added float_charge switch in batmon if it sees a BMS like '*PB2A16S*'

**Comment by personshelldon** [comment]
> So the voltage drop we see when BMS enters float mode comes from the BMS?

Hi! If correctly understand Your question than no. RFV (Float Voltage) it is just voltage that BMS requests from the inverter (via CAN or RS protocol). If inverter will ignore this request so no any voltage drop will be. For example, in my case when time of RFV comes - voltage slowly drops from RCV to RFV (battery can discharge a few percent during this time).

**Comment by Off-Grid-Garage** [comment]
> [@Off-Grid-Garage](https://github.com/Off-Grid-Garage) thanks for the explanation. So the voltage drop we see when BMS enters float mode comes from the BMS?

Yes, correct. 

> I guess it has an internal buck converter to step-down the voltage.

No, no, no, the BMS does not do that. It only sends the the signal to the charger (MPPT or Inverter) to lower the charging voltage. Otherwise the BMS would be huge to include a 10.000W buck converter.
 
> Does this work without talking to the solar charger?

No, it needs to have communication to the charger. Without comms, the charger has its own standard charging settings and does not react to any changes in the BMS. For example if the temperature in the battery rises, the charger does not know and keeps charging. So there is a potential risk of not having communication between the BMS and charger.

> Anyone knows which models has the float mode? looking at https://www.jkbms.com/products/active-balancer-bms/ , looks like the PB* models all have power inductors that could be part of the internal buck convert..

This website you stated is a fake site! The manufacturer site of the JK-BMS is http://www.jk-bms.com
I have links on my website to reliable sources where you can purchase these BMS. Also instructions how to set parameters and videos with tutorials... (sorry for the ad)

In general, the PBxx series BMS have the float mode functionality if flashed with the latest firmware.

**Comment by Off-Grid-Garage** [comment]
> I've added float_charge switch in batmon if it sees a BMS like '_PB2A16S_' based on [@personshelldon](https://github.com/personshelldon)'s patch

Current JK-BMS PBxx series models:
JK-PB1A16S10P (1A active balancer, 8-16s, 100A)
JK-PB1A16S15P (1A active balancer, 8-16s, 150A)
JK-PB2A16S15P (2A active balancer, 8-16s, 150A)
JK-PB2A16S20P (2A active balancer, 8-16s, 200A)

**Comment by fl4p** [comment]
@Off-Grid-Garage 
> No, no, no, the BMS does not do that. It only sends the the signal to the charger (MPPT or Inverter) to lower the charging voltage. Otherwise the BMS would be huge to include a 10.000W buck converter.

Not necessarily, as the buck would by bypassed if not in float mode and float current is usually much lower (10% ? ) of max charge current, so maybe 10A is enough. I see this is as a solution if no communication to the charger is available.


@roumano Looks like the python interpreter is too old, I'll see if I can pin a later version in the Dockerfile or port the code to be pre-3.12 compatible
